### DISCUSSION TITLE
Newbie question: bulk inserts; all fields

### DISCUSSION THREAD

**Discussion by tacman** [schranz-search/schranz-search#403] [Q&A] [answered]
Hi, I'm coming over from using the meilisearch-php library, and have some basic questions.

The meilisearch API support bulk adds, passing an array of json objects.  I don't see how do that with the $engine here.  Are there adapter-specific calls?  

Also, by default the meilisearch _stores_ all the fields you pass it in the addDocument() call, but you need to configure the searchable / filterable settings.  Is there a way in this library to store everything even if it's not defined in the schema?

Is the schema required to do searching, or is it just required when creating the index and indexing?  That is, if another tool creates the index can I simple call

```
$result = $this->engine->createSearchBuilder()
    ->addIndex('blog')
    ->addFilter(/* ... */)
    ->getResult();
```

without ever defining the schema?  Can I define the schema at run-time, rather than compile-time?  

Very cool package, thanks for releasing it!

**Comment by alexander-schranz** [comment]
@tacman

> Bulk Inserts

There is an open issue for this: https://github.com/schranz-search/schranz-search/issues/24 keep yourself uptodate by subscribe it.

> Schema

The schema is required as that defines how data is mapped from Index JSON <-> PHP Document. The Schema is very important as with SEAL we want make sure that you get always correct typed data back and so you only get the fields which you did define in your Schema, because Dynamic Schemas can make in complex projects a lot of issues if you allow Dynamic props. I know for prototyping its not ideal but for future of projects its better to think about your Document Structure types. Also have a look at: https://schranz-search.github.io/schranz-search/schema/index.html#best-practices

> Can a index be created external

It can work aslong as the external system create the Fields of the Index the same way as the SEAL would create it based on the Schema. Fields which are not defined in the SEAL mapping are ignored but not given back to you.

As mention above it was a Design decision to be very strict about Schema and types in SEAL to avoid Dynamic misusages and index fragmentations because of such things.

---

Thx for using Github Discussion and give SEAL a try let me know if you have any more questions.

**Reply by tacman** [reply]
Thanks for the quick and detailed reply.

The question about run-time indexes is because I don't know the data at compile-time.  For example, a user uploads a csv file, then tells us what fields they want indexed / filterable, then we dispatch an event to create the index and import the data.   So I'm mostly asking there about how to do it without defining the schema in /config/schemas, since that's hard-coded at compile-time.

**Reply by alexander-schranz** [reply]
Have a look at the `Standalone use` documentation part how to create Indexes via PHP directly instead of going over an static Index configuration of your Framework. Go to https://schranz-search.github.io/schranz-search/getting-started/index.html and click on `Standalone` and so you can create a "Dynamic Engine" class based on a Dynamic created Schema at runtime not at compile time of your dependency injection containers.

Still you may want to avoid 1 to 1 match between Index and your Domain Objects for Performance and Limitation reasons and example just put all searchable content into a `content` field see the Best practices Docs: https://schranz-search.github.io/schranz-search/schema/index.html#best-practices .

**Reply by alexander-schranz** [reply, marked as answer]
@tacman revisiting this. The SchemaLoader can be decorated to generate the Index Schema at runtime or add other possibilities.

See also: https://github.com/PHP-CMSIG/search/discussions/266#discussioncomment-12293374
